packages/dashboard/vite.config.ts

```diff
@@ -1,11 +1,58 @@
 /// <reference types="vitest" />
 
 import react from '@vitejs/plugin-react-swc';
-import { defineConfig } from 'vite';
+import { defineConfig, Plugin } from 'vite';
+
+import appConfig from './app-config.json';
+
+/**
+ * The goal of this plugin is to inject global variables to `index.html`, allowing
+ * a crude way to configure some variables after the bundle is built via `sed`.
+ *
+ * An example use case is building a dockerfile, because the domain in a dev or staging
+ * environment is typically different from prod, we would normally end up needing to build
+ * different images (which is really detrimental for staging). In such scenario, you can
+ * set the `rmfServerUrl` to a placeholder like `__RMF_SERVER_URL_PLACEHOLDER__` in
+ * `app-config.json`, then do a search and replace at the container entrypoint.
+ *
+ * The reason for doing this outside the app code is to avoid these variables from
+ * being modified by the bundler, and to reduce the chance that the search and replace modify
+ * unintended code.
+ */
+const injectGlobals: Plugin = {
+  name: 'injectRuntimeArgs',
+  transformIndexHtml: {
+    order: 'pre', // must be injected before the app
+    handler: () => {
+      const globals = {
+        RMF_SERVER_URL: appConfig.rmfServerUrl,
+        TRAJECTORY_SERVER_URL: appConfig.trajectoryServerUrl,
+      };
+      return [
+        {
+          tag: 'script',
+          injectTo: 'head',
+          children: Object.entries(globals)
+            .map(([k, v]) => `const ${k}='${v}'`)
+            .join(';'),
+        },
+      ];
+    },
+  },
+};
+
+function booleanToString(b: boolean | null | undefined) {
+  return b ? 'true' : 'false';
+}
 
 // https://vitejs.dev/config/
 export default defineConfig({
-  plugins: [react()],
+  define: {
```

Comment:
May I know what is different about `APP_CONFIG_AUTH_PROVIDER`, `APP_CONFIG_ENABLE_CUSTOM_TABS`, and `APP_CONFIG_ENABLE_ADMIN_TAB` from the other basic configurations in `app-config.json` like `helpLink`, `reportIssue`, such that they need to be defined here? It doesn't seem like we are using them for search-and-replace after the bundle is built too

Reply:
They need to be preprocessed before terser so that they can be tree shaken. vite will transform them to literals before passing to terser, if we reference them via a variable then terser can't analyze them.

Reply:
gotcha thanks for explaining